src/python/pants/option/errors.py

```diff
@@ -7,44 +7,89 @@
 class OptionsError(Exception):
   """An options system-related error."""
 
+# -----------------------------------------------------------------------
+# Option registration errors
+# -----------------------------------------------------------------------
 
 class RegistrationError(OptionsError):
   """An error at option registration time."""
 
-  def __init__(self, msg, scope, option):
-    super().__init__(
-      '{} [option {} in {}].'.format(msg, option,
-      'global scope' if scope == GLOBAL_SCOPE else 'scope {}'.format(scope)))
+  def __init__(self, scope: str, option: str, **msg_format_args) -> None:
+    scope_str = 'global scope' if scope == GLOBAL_SCOPE else f'scope {scope}'
+    if self.__doc__ is None:
+      raise ValueError(
+        "Invalid RegistrationError definition. Please specify the error message in the docstring."
+      )
+    docstring = self.__doc__.format(**msg_format_args)
```

Comment:
`mk_registration_error` wasn't understood by MyPy because it relied on runtime class creation, just like MyPy didn't understand `enum()` or `datatype()`. Here, we use the docstring to get the same effect of deduplication.